Don't create xiaomi_miio sensors if inital state is None

OGKevin opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
For the xiaomi_miio platform, when None is returned by the devices the following 2 assumptions can be made of its meaning:
1. The requested data is unsupported. e.g. asking a vacuum without a mop the status of its mop.
2. Glitches in the matrix that makes the device return None for some unknown reason. 

This PR introduces changes to handle these 2 assumptions.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #57474
- This PR is related to issue: https://github.com/home-assistant/core/issues/57474
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

OGKevin committed:
Accept None for some AirPurifier attributes.

Some AirPurifier attributes are know to return None every now and then, this
commit marks these events as accepted.

https://github.com/home-assistant/core/issues/57474

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

bieniu commented:
Honestly, I don't like this solution. I think we should ignore the value if it's `None`. Like here https://github.com/home-assistant/core/pull/57197

rytilahti commented:
I added a comment to the issue: https://github.com/home-assistant/core/issues/57474#issuecomment-941581989

OGKevin committed:
Mark the sensor as unavailable when its value is None.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Minor refactoring

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Mark binary_sensor and vacuum as unavailable if state is None.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Use _handle_coordinator_update to determine binary sensor state.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Use unknown state rather then setting entity to unavailable.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Return unknown state instead of None

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Merge remote-tracking branch 'origin/dev' into accept-none-for-air-purifier

OGKevin committed:
Return None in extra_state_attributes instad of state_attributes for vacuum.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Minor refactoring

Return None instead of Unknown state. The parent class translates None to
Unknown state.

Return extra_state_attributes when the state is unknown.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

MartinHjelmare commented:
There's a lint error.

OGKevin committed:
Silence lint for _parse_none.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

# Please enter the commit message for your changes. Lines starting
# with '#' will be ignored, and an empty message aborts the commit.
#
# On branch accept-none-for-air-purifier
# Your branch is up to date with 'origin/accept-none-for-air-purifier'.
#
# Changes to be committed:
#	modified:   homeassistant/components/xiaomi_miio/device.py
#

OGKevin commented:
Lint should be good now 🤞🏾.

@bieniu @rytilahti any feedback?

OGKevin committed:
Minor refactoring of pylint disables.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

jbouwh reviewed (changes requested):
`None` should not be accepted as a valid state.

OGKevin commented:
@jbouwh There is no other option atm the to accept that this attribute is returning `None` every now and then. `None` should indeed not be accepted as normal behavior. But in the case for these Air Purifiers there is no other option. Unless someone dedicates their time in investigating why these attributes are returning `None` every now and then.

This PR is "Accepting" the `None` by not generating a warning that an unexpected `None` has been returned.

A better solution would be is to investigate why these attributes return None every now and then and fix it. But I cant do that as I don't have such device. Check the linked issue for more context.

E.g. for Vacuums, a separate PR will be created to not create the sensors in the first place.

I'm open for suggestions on how to properly handle None for "unreliable" attributes.

OGKevin committed:
Move pylint disable statements

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

OGKevin commented:
Please keep in mind that this is a bug fix and i want to get this fixed ASAP before the Nov release. I still need to fix the vacuum sensors and that work will branch off this fix.

OGKevin commented:
> My advice is to ignore None-states if the sensor is supported. A temporary None should just mean there is no update. To indicate a sensor is not available we can make in un-available, but only if we know what the behavior is

When `None` is returned as state, the entity will be marked as "Unknown State" by HA. 

@MartinHjelmare is against marking it as Unavailable, as this means that the device is unavailable, which is not true. The device is available, the state is just temp unknown due to the "known" None that is returned every now and then.

> I understand, but the way to approach this would be perhaps in the backend library, not here.

I agree, but i do not have such device in my possession to do proper testing and supply a proper fix to the lib. So a solution in HA is the quick win for now.

>  The integration now updates data every 15 seconds so if the value None appears, for 15 seconds the entity should use the old state and then get the new one.

So you are saying, if these attributes that are know to return None every now and then, to ignore the None and keep the previous state? And only update once a proper value is returned? @MartinHjelmare what do you think of this approach? Is it acceptable?

OGKevin committed:
Ignore None state when allowed.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Log warning after 3 consecutive allowed None's.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Remove redundant else statement.

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

OGKevin committed:
Add docs on ignoring updates

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Filter out None attributes from state_attributes

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

rytilahti commented:
After doing some thinking on this, I'm fine with it if other reviewers agree, even when I don't like the solution. The reasons not to like this:
* Adding extra logic (counters, extra parameters to sensor definitions) will make the code more complicated by piling more logical decisions to the integration and will increase the maintenance burden
* I'm not (still) seeing what value the user is getting here (compared to simply suppressing on error cases)

To my understanding, there are two cases when the sensor state could be `None`:
1. The sensor returns `None` during the initial update. This happens most likely because the feature is not supported by the device
2. The sensor value "flaps" from a valid value to `None`. This is likely caused by some hardware failure

An alternative approach to fix the issue:
1. Log a warning on `None`s during initial update to let the user know that they should report incorrectly added entity (case "non-supported sensor"). This might result to some false positives, but I find it better than suppressing the fact that the device is not behaving as expected.
2. If the state unexpectedly changes to `None` (case "flapping sensor"), debug log it and use the existing value. If needed, we can try to find a solution for this case-by-case basis, but my personal opinion is that when other information is correctly reported by the device, the backend library should not interfere (or do forced updates or anything like that).

This way we can avoid adding new logic and sensor parameters while still avoiding spamming the log on occasional sensor failures. 

Opinions? Like @OGKevin, I'd really like to see this fixed by the next release when not even for a point release.

OGKevin committed:
Fix comment syntax.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Merge remote-tracking branch 'origin/dev' into accept-none-for-air-purifier

OGKevin commented:
Alright, ill make the change then.

OGKevin committed:
Merge branch 'dev' into accept-none-for-air-purifier

jbouwh commented:
> Alright, ill make the change then.

Where is this code coming from if I may ask?

OGKevin committed:
Refactored xiaomi sensors to not create if initial state is none.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin committed:
Include initial state check in all sensors

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>

OGKevin commented:
> Where is this code coming from if I may ask?

I don't understand your question :P can you elaborate?

jbouwh commented:
> > Where is this code coming from if I may ask?
> 
> I don't understand your question :P can you elaborate?

Forget my remark, I was confused.

bieniu commented:
@OGKevin I took your code and changed it a bit 😉 I think we should go in that direction. Please have a look https://github.com/home-assistant/core/compare/dev...bieniu:xiaomi_miio/none_values

OGKevin commented:
Can you make that diff against my branch its kinda hard to keep track of the changes. But so far what i can see is that you ignore unexpected value types in `_extract_value_from_attribute` which i personally don't like. I like to be specific/explicit.

And you log info instead of debug when the initial sensors is not created which is a good idea 👍🏾.

The rest seems roughly the same.

So why ignore unexpected value types?

OGKevin committed:
Add info log if sensor is not supported.

Signed-off-by: Kevin Hellemun <17928966+OGKevin@users.noreply.github.com>
co-authored-by: Maciej Bieniek <bieniu@users.noreply.github.com>

OGKevin commented:
@MartinHjelmare can i get some 👀

bieniu commented:
> Can you make that diff against my branch its kinda hard to keep track of the changes.

Sure. https://github.com/bieniu/home-assistant-core/compare/accept-none-for-air-purifier...bieniu:xiaomi_miio/none_values

> The rest seems roughly the same.

IMO the code is different.

OGKevin commented:
>  Can’t automatically merge. Don’t worry, you can still create the pull request. 

Makes the diff confusing as it seems we both authored the same change :P hence I asked. Expected you to resolve the conflicts hehe. 

> IMO the code is different.

In what aspect? 

You have 
```python
       parent_key_data = getattr(coordinator.data, description.parent_key)
        if getattr(parent_key_data, description.key, None) is None:
            _LOGGER.info(
                "The %s device does not support the %s sensor",
                config_entry.data[CONF_MODEL],
                description.key,
            )
            continue
```
I do it, in my eyes, a little more elegantly by reusing code: 
```python
sensor = XiaomiGenericSensor(
                    f"{config_entry.title} {description.name}",
                    device,
                    config_entry,
                    f"{sensor_name}_{config_entry.unique_id}",
                    hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
                    description,
                )

                # Don't create the sensor if its initial state is None.
                # Initial update is done in at coordinator creation
                if sensor.native_value is not None:
                    entities.append(sensor)
```
as `native_value` under the hood already calls coordinator.data. with the right parent key _if_ it is set.

You also forgot to set it for the other platforms, not sure if it was intended or not.

But the rest i think both our code achieves the same with just diff implementation opinions. I like to be explicit, hence i left the check for 

```
        if isinstance(value, float):
            return value
        if isinstance(value, int):
            return value
```

As these are confirmed returned types that are handled accordingly.

But in your version you removed that, along with the warning log of unexpected type.

Just look at the reported issues, by being explicit:
1. we now have a bunch of report of models that have silently been working fine but no one bother to add them to the supported list
2. we can now build a list of features that models support. 

But you are a maintainer so In the end it is your call and i will respect it, I invite you to make the changes as you please and push them to this PR.

bieniu commented:
> ```python
> sensor = XiaomiGenericSensor(
>                     f"{config_entry.title} {description.name}",
>                     device,
>                     config_entry,
>                     f"{sensor_name}_{config_entry.unique_id}",
>                     hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
>                     description,
>                 )
> 
>                 # Don't create the sensor if its initial state is None.
>                 # Initial update is done in at coordinator creation
>                 if sensor.native_value is not None:
>                     entities.append(sensor)
> ```

We don't want to create the `XiaomiGenericSensor` object if we can not use it later if the value is` None`.

> You also forgot to set it for the other platforms, not sure if it was intended or not.

Yes, this is intentional, we don't have reports about `None` values related to other platforms.

> ```
>         if isinstance(value, float):
>             return value
>         if isinstance(value, int):
>             return value
> ```
> 
> As these are confirmed returned types that are handled accordingly.
> 
> But in your version you removed that, along with the warning log of unexpected type.

IMO this is not necessary, only `None` is an unexpected value.

OGKevin commented:
>      We don't want to create the XiaomiGenericSensor object if we can not use it later if the value is None.

why not? Isn't it more elegant to re-use the internal code instead of copy pasting it? When the code changes you have to go and change it in both/all places.

> Yes, this is intentional, we don't have reports about None values related to other platforms.

Is [this](https://github.com/home-assistant/core/issues/57474#issuecomment-940217322) not unsupported? Or also just glitch?

> IMO this is not necessary, only None is an unexpected value. I think that logging unexpected None is ok but with debug level.

This is where we disagree :). So up to you. The None in my code is also debug lvl. The warning remained if its an unexpected type. 

But again, just push your changes in this PR or make it a clean/mergable diff against mine. This has taken enough time already.

bieniu commented:
> why not? Isn't it more elegant to re-use the internal code instead of copy pasting it? When the code changes you have to go and change it in both/all places.

IMO it's not elegant to create objects that may not be used.

> Is [this](https://github.com/home-assistant/core/issues/57474#issuecomment-940217322) not unsupported? Or also just glitch?

Glitch

> The warning remained if its an unexpected type

When can this happen?

> But again, just push your changes in this PR or make it a clean/mergable diff against mine. This has taken enough time already.

Sorry, but this is your PR and you have to decide what to do with it and how to change it. I can open a new PR with my solution but I need someone to test the vacuum related part as I don' have such a device.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/binary_sensor.py:
```diff
@@ -151,17 +163,30 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
-                    hass.data[DOMAIN][config_entry.entry_id][KEY_COORDINATOR],
-                    description,
-                )
+            sensor = XiaomiGenericBinarySensor(
```
It's nice if we can avoid creating the entity if the data value is None and we won't add the entity.

MartinHjelmare commented on homeassistant/components/xiaomi_miio/sensor.py:
```diff
@@ -617,20 +625,33 @@ async def async_setup_entry(hass, config_entry, async_add_entities):
+                        "Sensor %s has not been created due to its initial value being None/Unknown.",
+                        sensor.name,
+                    )
+                    _LOGGER.info(
```
Please only log once.

balloob commented on homeassistant/components/xiaomi_miio/binary_sensor.py:
```diff
@@ -116,17 +120,25 @@ def _setup_vacuum_sensors(hass, config_entry, async_add_entities):
+            entities.append(sensor)
+        else:
+            _LOGGER.debug(
+                "Sensor %s has not been created due to its initial value being None/Unknown.",
```
```suggestion
                "Sensor %s has not been created due to its initial value being None/Unknown",
```

balloob commented:
Removed second opinion wanted. Agree with that we just not add it when it’s None, a next reboot will cause it to be created which is ok.

OGKevin commented:
> Sorry, but this is your PR and you have to decide what to do with it and how to change it. I can open a new PR with my solution but I need someone to test the vacuum related part as I don' have such a device.

I'm tired of this back and forth, it has been 20 days. I'm pretty sure that my next push will again have feedback that goes against my opinions/practices etc. I've been outvoted and that's fine. I just want to get this merged, I'm slightly pissed that on 31st of Oct there is finally some sort of concessions in this discussion for a release of November. Seeing that you won the vote it makes the most sense for you to make the change as you know what you want to do. 

So, if you pushing the change or opening a new PR etc is guaranteed to go faster then that's the best approach. I would be happy to test on my vacuum.

OGKevin commented:
Closing in favour of https://github.com/home-assistant/core/pull/58880#issuecomment-956110947
